app/models/gem_typo.rb

```diff
@@ -3,43 +3,33 @@
 class GemTypo
   attr_reader :protected_gem
 
-  include Gem::Text
-
-  DOWNLOADS_THRESHOLD = 10_000_000
-  SIZE_THRESHOLD = 4
-
   def initialize(rubygem_name)
-    @rubygem_name       = rubygem_name.downcase
-    @distance_threshold = distance_threshold
+    @rubygem_name = rubygem_name
   end
 
   def protected_typo?
-    return false if @rubygem_name.size < GemTypo::SIZE_THRESHOLD
+    return false if GemTypoException.where("upper(name) = upper(?)", @rubygem_name).any?
 
-    gem_typo_exceptions = GemTypoException.all.pluck(:name)
-    return false if gem_typo_exceptions.include?(@rubygem_name)
+    return false if published_exact_name_matches.any?
 
-    protected_gems.each do |protected_gem|
-      distance = levenshtein_distance(@rubygem_name, protected_gem)
-      if distance <= @distance_threshold
-        @protected_gem = protected_gem
-        return true
-      end
-    end
+    match = matched_protected_gem_names.select(:name).first
```

Comment:
Because of using `first` here, rails is adding `order by` and it is not using the intended index:
```
rubygems_development=# explain analyze SELECT "rubygems"."name" FROM "rubygems" WHERE "rubygems"."indexed" = true AND (regexp_replace(upper(name), '[_-]', '', 'g') = r
egexp_replace(upper('r__d-oc-js--on---'), '[_-]', '', 'g')) ORDER BY "rubygems"."id" ASC limit 1;
                                                               QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.42..15.43 rows=1 width=18) (actual time=310.622..310.623 rows=1 loops=1)
   ->  Index Scan using rubygems_pkey on rubygems  (cost=0.42..12038.09 rows=802 width=18) (actual time=310.620..310.620 rows=1 loops=1)
         Filter: (indexed AND (regexp_replace(upper((name)::text), '[_-]'::text, ''::text, 'g'::text) = 'RDOCJSON'::text))
         Rows Removed by Filter: 172345
 Planning time: 0.340 ms
 Execution time: 310.659 ms
(6 rows)

rubygems_development=# explain analyze SELECT "rubygems"."name" FROM "rubygems" WHERE "rubygems"."indexed" = true AND (upper(name) != upper('r__d-oc-json---') and
regexp_replace(upper(name), '[_-]', '', 'g') = regexp_replace(upper('r__d-oc-js--on---'), '[_-]', '', 'g'));
                                                             QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------
 Bitmap Heap Scan on rubygems  (cost=23.09..1414.05 rows=802 width=14) (actual time=0.048..0.050 rows=1 loops=1)
   Recheck Cond: (regexp_replace(upper((name)::text), '[_-]'::text, ''::text, 'g'::text) = 'RDOCJSON'::text)
   Filter: (indexed AND (upper((name)::text) <> 'R__D-OC-JSON---'::text))
   Heap Blocks: exact=1
   ->  Bitmap Index Scan on dashunderscore_typos_idx  (cost=0.00..22.89 rows=862 width=0) (actual time=0.033..0.033 rows=1 loops=1)
         Index Cond: (regexp_replace(upper((name)::text), '[_-]'::text, ''::text, 'g'::text) = 'RDOCJSON'::text)
 Planning time: 0.299 ms
 Execution time: 0.096 ms
(8 rows)

```